src/background.ts

```diff
@@ -68,18 +67,32 @@ async function addInjectionListener(context: CacheContext) {
 	const allDomains = Object.values<string[]>(injectionDomains as any).flat();
 
 	// note: This is a closure over injectionDomains
-	const injectScript = (details: WebNavigation.OnDOMContentLoadedDetailsType) => {
+	const injectScript = (tabId: number, tabUrl: string) => {
 		void scripting.executeScript({
-			target: { tabId: details.tabId },
+			target: { tabId: tabId },
 			// injectImmediately: true,
-			func: getInjectionFn(details.url, injectionDomains),
-			args: [details.url, GKDotDevUrl],
+			func: getInjectionFn(tabUrl, injectionDomains),
+			args: [tabUrl, GKDotDevUrl],
 		});
 	};
 
-	webNavigation.onDOMContentLoaded.addListener(injectScript, {
+	webNavigation.onDOMContentLoaded.addListener(details => injectScript(details.tabId, details.url), {
 		url: allDomains.map(domain => ({ hostContains: domain })),
 	});
+
+	// Immediately inject into the currently open compatible tabs. This is needed because when the background
+	// script is idle, its event listeners are not active. Opening a compatible tab will cause the background
+	// script to awaken and setup the event listeners again, but the tab will load before that happens.
```

Comment:
This is a clever workaround but I'm wondering if we can make it less hacky.

I didn't verify this, but it seems the reason the background page already awakens on a compatible tab is because of the `webNavigation.onHistoryStateUpdated` listener above. If it weren't for this listener, the background script wouldn't activate on new tabs at all.

Ideally we'd have `webNavigation.onDOMContentLoaded` be executed globally like the rest of the listeners, so it will awaken the background script. When I initially wrote this, I couldn't think of another way to do it without using session storage and adding the `storage` permission. I'm thinking there is another way to do it without storage, but I'd need to sit on it more.

If we do not want to move `webNavigation.onDOMContentLoaded` outside of `addInjectionListener`, there might be a "less hacky" way to implement this by adding another `onDOMContentLoaded` globally, so we can ensure the background page always awakens in case we decide to remove the `onHistoryStateUpdated` listener one day.

Reply:
I didn't realize you already added the `storage` permission. In that case, you can make `webNavigation.onDOMContentLoaded` a global listener that reads from session storage to get the domains to inject. This also means we'd have to remove the filter parameter for the callback, but it doesn't impact efficiency at all since `webNavigation.onHistoryStateUpdated` is already firing on any URL changes.

Furthermore I don't think it's necessary to inject into current tabs anymore, but it's a neat feature if you want it to work on tabs opened before the extension has loaded. But I'd be careful to ensure the scripts are only injected once; I think it'll cause problems if they're injected more than once.